### PR TITLE
Use NewTagMap in place of ExtensionMap and UnknownFieldMap.

### PR DESCRIPTION
**Pull request by swankjesse**
There's a lot of fall out from this that can be cleaned up in follow-up. In particular,
we can tear down some Storage and TagBinding code.

### CODE REVIEW

File: wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java [hunk 1]
```diff
@@ -78,46 +77,58 @@ public MessageTypeAdapter(Wire wire, Gson gson, TypeToken<M> type) {
     }
 
     out.beginObject();
-    for (TagBinding<M, Message.Builder<M>> tagBinding
-        : messageAdapter.tagBindingsForMessage(message).values()) {
+    for (TagBinding<M, Message.Builder<M>> tagBinding : messageAdapter.tagBindings().values()) {
       Object value = tagBinding.get(message);
-      if (value == null) {
+      if (value == null || tagBinding instanceof ExtensionTagBinding) {
```
Comment (swankjesse) [inline wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.java]: (these instanceofs are gross; they prevent double-counting)

File: wire-schema/src/test/java/com/squareup/wire/schema/SchemaWireAdapterTest.java [hunk 2]
```diff
@@ -170,9 +170,9 @@
         .buildWireAdapter("Message");
     ImmutableMap<String, Object> expected = ImmutableMap.<String, Object>of(
         "a", ImmutableList.of(601, 701));
-    ByteString packedEncoded = ByteString.decodeHex("d005d904d005bd05");
+    ByteString packedEncoded = ByteString.decodeHex("d20504d904bd05");
```
Comment (swankjesse) [inline wire-schema/src/test/java/com/squareup/wire/schema/SchemaWireAdapterTest.java]: (these were flipped by mistake)

File: wire-runtime/src/main/java/com/squareup/wire/TagMap.java [hunk 3]
```diff
@@ -1,139 +0,0 @@
...
-import java.util.TreeSet;
-
-/**
- * A read-only subset of Map&lt;Integer, T&gt; optimized for the case where the set of tags is
- * reasonably small or dense (defined in terms of the ratio of the number of active entries to the
- * max tag value).
- *
- * @param <T> the output type of the map.
- */
-abstract class TagMap<T> {
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/TagMap.java]: This name is now up for grabs!

File: wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java [hunk 4]
```diff
@@ -37,17 +35,15 @@ protected ExtendableMessage() {
    */
   protected void setBuilder(ExtendableBuilder<T, ?> builder) {
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java]: Remove whole method.

File: wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java [hunk 5]
```diff
@@ -83,7 +78,6 @@ protected int extensionsHashCode() {
       B extends ExtendableBuilder<T, B>> extends Builder<T> {
 
     @SuppressWarnings("unchecked")
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java]: Remove

File: wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java [hunk 6]
```diff
@@ -103,17 +94,24 @@ protected ExtendableBuilder(Class<B> selfType, ExtendableMessage<T> message) {
...
     }
 
     /**
      * Sets the value of {@code extension} on this builder to {@code value}.
      */
     public <E> B setExtension(Extension<T, E> extension, E value) {
-      if (extensionMap == null) {
-        extensionMap = new ExtensionMap<T>(extension, value);
+      if (tagMap == null) {
+        tagMap = new NewTagMap();
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java]: `new NewTagTagMapMap()`

Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java]: `tagTagMapMap = new NewTagTagMapMap();`, but I digress...

File: wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java (hunk 5 above)
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java]: Oh wow, that's a weird edit error. Fixed.

File: wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java (hunk 4 above)
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/ExtendableMessage.java]: Fixed.

File: wire-runtime/src/main/java/com/squareup/wire/NewTagMap.java [hunk 7]
```diff
@@ -73,52 +89,175 @@ public void add(Extension<?, ?> extension, Object value) {
...
+    } else if (list.size() == 1) {
+      return list.get(0);
+    } else {
+      throw new IllegalArgumentException(
+          "found multiple values for non-repeated extension " + targetExtension);
+    }
   }
 
-  private Object decode(Extension<?, ?> target, Extension<?, ?> source, Object value) {
+  private void transcode(List<Object> list, Extension<?, ?> sourceExtension,
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/NewTagMap.java]: Oooo transcode. Nice name.

**Comment by JakeWharton** [comment]
LGTM!

File: wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java [hunk 8]
```diff
@@ -33,7 +33,6 @@
 import static com.squareup.wire.Message.Builder;
 
 final class RuntimeMessageAdapter<M extends Message> extends WireAdapter<M> {
-  private final Wire wire;
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java]: Best part of this change, right here!

**Pull request merged**